arangod/Replication2/ReplicatedLog/LogLeader.cpp

```diff
@@ -0,0 +1,945 @@
+        << "in appendEntries to follower " << follower._impl->getParticipantId()
+        << ", result future has neither value nor exception.";
+    TRI_ASSERT(false);
+    FATAL_ERROR_EXIT();
+  }
+
+  // try sending the next batch
+  return std::make_pair(prepareAppendEntries(), std::move(toBeResolved));
+}
+
+auto replicated_log::LogLeader::GuardedLeaderData::getInternalLogIterator(LogIndex firstIdx) const
+    -> std::unique_ptr<PersistedLogIterator> {
+  auto const endIdx = _inMemoryLog.getLastTermIndexPair().index + 1;
+  TRI_ASSERT(firstIdx <= endIdx);
+  return _inMemoryLog.getInternalIteratorFrom(firstIdx);
+}
+
+auto replicated_log::LogLeader::GuardedLeaderData::getCommittedLogIterator(LogIndex firstIndex) const
+    -> std::unique_ptr<LogIterator> {
+  auto const endIdx = _inMemoryLog.getNextIndex();
+  TRI_ASSERT(firstIndex < endIdx);
+  // return an iterator for the range [firstIndex, _commitIndex + 1)
+  return _inMemoryLog.getIteratorRange(firstIndex, _commitIndex + 1);
+}
+
+auto replicated_log::LogLeader::GuardedLeaderData::checkCommitIndex() -> ResolvedPromiseSet {
+  auto const quorum_size = _self._config.writeConcern;
+
+  std::vector<std::pair<LogIndex, ParticipantId>> indexes;
+  indexes.reserve(_follower.size());
+  for (auto const& follower : _follower) {
+    // The lastAckedEntry is the last index/term pair that we sent that this
+    // follower acknowledged - means we sent it. And we must not have entries
+    // in our log with a term newer than currentTerm, which could have been
+    // sent to a follower.
+    auto const& lastAckedEntry = follower.lastAckedEntry;
+    TRI_ASSERT(lastAckedEntry.term <= this->_self._currentTerm);
+    // We must never commit log entries for older terms, as these could still be
+    // overwritten later if a leader takes over that holds an entry with the
+    // same index, but with a newer term than that entry has.
+    // For more details and an example see the Raft paper, specifically on
+    // page 9 both subsection "5.4.2 Committing entries from previous terms" and
+    // figure 8.
+    // We may only commit these if we've written an entry in our current term.
+    // This also includes log entries persisted on this server, i.e. our
+    // LocalFollower is no exception.
+    if (lastAckedEntry.term == this->_self._currentTerm) {
+      indexes.emplace_back(lastAckedEntry.index, follower._impl->getParticipantId());
+    } else {
+      LOG_CTX("54869", TRACE, _self._logContext)
+          << "Will ignore follower "
+          << follower._impl->getParticipantId() << " in the following commit index check, as its last log entry (index "
+          << lastAckedEntry.index << ") is of term " << lastAckedEntry.term
+          << ", but we're in term " << _self._currentTerm << ".";
+    }
+  }
+
+  LOG_CTX("a2d04", TRACE, _self._logContext) << "checking commit index on set " << indexes;
+
+  if (quorum_size <= 0 || quorum_size > indexes.size()) {
```

Comment:
`quorum_size` is unsigned, so it cannot be < 0.
```suggestion
  if (quorum_size == 0 || quorum_size > indexes.size()) {
```